probe-rs/src/architecture/arm/core/armv7a.rs

```diff
@@ -297,29 +274,271 @@ impl<'probe> Armv7a<'probe> {
+            return Err(ArmError::Timeout);
+        }
+        // Wait a bit before polling again.
+        std::thread::sleep(Duration::from_millis(1));
     }
 
-    fn core_halted(&mut self) -> Result<bool, Error> {
-        let address = Dbgdscr::get_mmio_address_from_base(self.base_address)?;
-        let dbgdscr = Dbgdscr(self.memory.read_word_32(address)?);
+    Ok(())
+}
+
+/// Return whether or not the core is halted.
+fn core_halted(memory: &mut dyn ArmMemoryInterface, base_address: u64) -> Result<bool, ArmError> {
+    let address = Dbgdscr::get_mmio_address_from_base(base_address)?;
+    let dbgdscr = Dbgdscr(memory.read_word_32(address)?);
+
+    Ok(dbgdscr.halted())
+}
+
+/// Set and enable a specific breakpoint. If the breakpoint is in use, it
+/// will be cleared.
+pub(crate) fn set_hw_breakpoint(
+    memory: &mut dyn ArmMemoryInterface,
+    base_address: u64,
+    bp_unit_index: usize,
+    addr: u32,
+) -> Result<(), ArmError> {
+    let bp_value_addr = Dbgbvr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+    let bp_control_addr = Dbgbcr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+    let mut bp_control = Dbgbcr(0);
+
+    // Breakpoint type - address match
+    bp_control.set_bt(0b0000);
+    // Match on all modes
+    bp_control.set_hmc(true);
+    bp_control.set_pmc(0b11);
+    // Match on all bytes
+    bp_control.set_bas(0b1111);
+    // Enable
+    bp_control.set_e(true);
+
+    memory.write_word_32(bp_value_addr, addr)?;
+    memory.write_word_32(bp_control_addr, bp_control.into())?;
+
+    Ok(())
+}
 
-        Ok(dbgdscr.halted())
+/// If a specified breakpoint is set, disable it and clear it.
+pub(crate) fn clear_hw_breakpoint(
+    memory: &mut dyn ArmMemoryInterface,
+    base_address: u64,
+    bp_unit_index: usize,
+) -> Result<(), ArmError> {
+    let bp_value_addr = Dbgbvr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+    let bp_control_addr = Dbgbcr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+
+    memory.write_word_32(bp_value_addr, 0)?;
+    memory.write_word_32(bp_control_addr, 0)?;
+    Ok(())
+}
+
+/// Get a specific hardware breakpoint. If the breakpoint is not set, return `None`.
+pub(crate) fn get_hw_breakpoint(
+    memory: &mut dyn ArmMemoryInterface,
+    base_address: u64,
+    bp_unit_index: usize,
+) -> Result<Option<u32>, ArmError> {
+    let bp_value_addr = Dbgbvr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+    let bp_value = memory.read_word_32(bp_value_addr)?;
+
+    let bp_control_addr = Dbgbcr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+    let bp_control = Dbgbcr(memory.read_word_32(bp_control_addr)?);
+
+    Ok(if bp_control.e() { Some(bp_value) } else { None })
```

Comment:
I wonder why clippy doesn't want this to be `bp_control.e().then_some(bp_value)`. Regardless, I suspect this is a bit of copy-paste code, but I'd prefer not reading Dbgbvr and memory unless the breakpoint is actually set. WDYT?